daemons/base.py

```diff
@@ -231,67 +265,64 @@ def parse_params(self, params):
-            return web.json_response(
-                {"jsonrpc": "2.0", "error": {"code": -32601, "message": "Procedure not found."}, "id": id}
-            )
+        return id, method, args, kwargs, error
+
+    async def get_exec_method(self, cmd, id, req_method):
+        exec_method, custom = (
+            (self.supported_methods[req_method], True)
+            if req_method in self.supported_methods
+            else (getattr(cmd, req_method, JsonResponse(code=-32601, error="Procedure not found", id=id)), False)
+        )
+        error = exec_method if isinstance(exec_method, JsonResponse) else None
+        return exec_method, custom, error
+
+    async def get_exec_result(self, req_method, req_args, req_kwargs, exec_method, custom, **kwargs):
+        xpub = req_kwargs.pop("xpub", None)
+        if custom:
+            exec_method = functools.partial(exec_method, wallet=xpub)
+        else:
+            if self.NEW_ELECTRUM and self.electrum.commands.known_commands[req_method].requires_wallet:
+                wallet, config = kwargs.get("wallet"), kwargs.get("config")
+                cmd_name = self.electrum.commands.known_commands[req_method].name
+                need_path = cmd_name in ["create", "restore"]
+                path = wallet.storage.path if wallet else (config.get_wallet_path() if need_path else None)
+                exec_method = functools.partial(exec_method, wallet_path=path)
+        result = exec_method(*req_args, **req_kwargs)
+        return await result if inspect.isawaitable(result) else result
+
+    async def _get_wallet(self, id, req_method, req_kwargs):
+        wallet = cmd = config = error = None
+        xpub = req_kwargs.get("xpub", None)
         try:
             wallet, cmd, config = await self.load_wallet(xpub)
         except Exception:
-            if method not in self.supported_methods:
-                return web.json_response(
-                    {"jsonrpc": "2.0", "error": {"code": -32601, "message": "Error loading wallet"}, "id": id}
-                )
-        custom = False
-        if method in self.supported_methods:
-            exec_method = self.supported_methods[method]
-            custom = True
-        else:
-            try:
-                exec_method = getattr(cmd, method)
-            except AttributeError:
-                return web.json_response(
-                    {"jsonrpc": "2.0", "error": {"code": -32601, "message": "Procedure not found."}, "id": id}
-                )
+            if req_method not in self.supported_methods:
+                error = JsonResponse(code=-32601, error="Error loading wallet", id=id)
+        return wallet, cmd, config, error
+
+    @authenticate
+    async def handle_request(self, request):
+        id, req_method, req_args, req_kwargs, error = await self.get_handle_request_params(request)
+        if error:
+            return error.send()
+        wallet, cmd, config, error = await self._get_wallet(id, req_method, req_kwargs)
```

Comment:
I suggest to pop xpub early and just pass it to other functions later. (Avoids duplicate `.get("xpub", None)` etc.)
```suggestion
        xpub = req_kwargs.pop("xpub", None)
        wallet, cmd, config, error = await self._get_wallet(id, req_method, xpub)
```